File: esn_tutorial_py/esn_tutorial_py/robot_action_server.py
#!/usr/bin/env python3
import time

import rclpy
from rclpy.node import Node
from rclpy.action import ActionServer, GoalResponse, CancelResponse

from geometry_msgs.msg import Twist
from esn_msgs.action import PickObject

# import numpy as np

class RobotActionServer(Node):
    def __init__(self):
        super().__init__('robot_action_server_node')

        # Parametri
        self.declare_parameter('motion_duration', 2.0)
        self.motion_duration = self.get_parameter('motion_duration').value

        self.declare_parameter('initial_x', 0.0)
        self.initial_x = self.get_parameter('initial_x').value

        self.declare_parameter('initial_y', 0.0)
        self.initial_y = self.get_parameter('initial_y').value

        self.declare_parameter('action_name', 'pick_object')
        action_name = self.get_parameter('action_name').value

        # Publisher
        self.cmd_pub = self.create_publisher(Twist, '/cmd_vel', 10)

        # Action server
        self._action_server = ActionServer(
            self,
            PickObject,
            action_name,
            execute_callback=self.__execute_callback,
            goal_callback=self.__goal_callback,
            cancel_callback=self.__cancel_callback,
        )
        self.robot_busy = False

    # ---- GOAL MANAGEMENT ----
    def __goal_callback(self, goal_request: PickObject.Goal):
        self.get_logger().info(
            f"Received goal pose [{goal_request.pose.header.frame_id}] "
            f"pos=({goal_request.pose.pose.position.x:.2f}, "
            f"{goal_request.pose.pose.position.y:.2f})"
        )

        # q = goal_request.pose.orientation
        # orientation_vec = np.array([q.x, q.y, q.z, q.w])
        # if not np.isclose(np.linalg.norm(orientation_vec), 1.0, atol=1e-6):
        #     self.get_logger().warning("Rejecting goal: quaternion is not normalized.")
        #     return GoalResponse.REJECT

        if self.robot_busy:
            self.get_logger().warning('Rejecting goal: robot is busy.')
            return GoalResponse.REJECT

        self.get_logger().info('Accepting goal.')
        self.robot_busy = True
        return GoalResponse.ACCEPT

    def __cancel_callback(self, goal_handle) -> CancelResponse:
        self.get_logger().info("Cancel request received")
        return CancelResponse.ACCEPT

    # ---- EXECUTION ----
    def __execute_callback(self, goal_handle):
        self.get_logger().info("Executing goal...")

        goal: PickObject.Goal = goal_handle.request
        target_x = float(goal.pose.pose.position.x)
        target_y = float(goal.pose.pose.position.y)

        dx = target_x - self.initial_x
        dy = target_y - self.initial_y

        T = self.motion_duration if self.motion_duration > 1e-6 else 1e-6
        vx = dx / T
        vy = dy / T

        self.get_logger().info(
            f"Computed velocity: vx={vx:.3f}, vy={vy:.3f} to reach in {T:.2f}s"
        )

        rate_hz = 10.0
        dt = 1.0 / rate_hz
        period = dt

        t = 0.0
        cmd = Twist()
        stop = Twist()

        feedback = PickObject.Feedback()
        result = PickObject.Result()

        while t < T and rclpy.ok():
            if goal_handle.is_cancel_requested:
                self.cmd_pub.publish(stop)
                feedback.progress = float(t / T)
                goal_handle.publish_feedback(feedback)

                result.success = False
                goal_handle.canceled()
                self.get_logger().warn(
                    f"Goal canceled at {feedback.progress * 100:.0f}%"
                )
                self.robot_busy = False
                return result

            cmd.linear.x = vx
            cmd.linear.y = vy
            self.cmd_pub.publish(cmd)

            feedback.progress = float(t / T)
            goal_handle.publish_feedback(feedback)

            time.sleep(period)
            t += dt
        # fermo sempre il robot
        self.cmd_pub.publish(stop)
        self.robot_busy = False
        
        if rclpy.ok():
            result.success = True
            goal_handle.succeed()
            self.get_logger().info("Goal succeeded, reached target pose")

            # aggiorno la posizione iniziale
            self.initial_x = target_x
            self.initial_y = target_y

        return result

File: esn_tutorial/src/robot_action_server.cpp
#include <esn_tutorial/robot_action_server.hpp>
#include <functional>
#include <thread>

namespace esn {

RobotAction::RobotAction() : Node("robot_action_node")
{
    this->declare_parameter("motion_duration", 2.0);
    this->motion_duration_ = this->get_parameter("motion_duration").as_double();

    this->declare_parameter("initial_x", 0.0);
    this->initial_x_ = this->get_parameter("initial_x").as_double();

    this->declare_parameter("initial_y", 0.0);
    this->initial_y_ = this->get_parameter("initial_y").as_double();

    this->declare_parameter("action_name", "take_object");
    std::string action_name = this->get_parameter("action_name").as_string();

    this->cmd_vel_pub_ = this->create_publisher<geometry_msgs::msg::Twist>("/cmd_vel", 10);

    this->action_server_ = rclcpp_action::create_server<esn_msgs::action::PickObject>(
        this,
        action_name,
        std::bind(&RobotAction::handle_goal, this, std::placeholders::_1, std::placeholders::_2),
        std::bind(&RobotAction::handle_cancel, this, std::placeholders::_1),
        std::bind(&RobotAction::handle_accepted, this, std::placeholders::_1)
        );
}

rclcpp_action::GoalResponse RobotAction::handle_goal(
    const rclcpp_action::GoalUUID &,
    std::shared_ptr<const esn_msgs::action::PickObject::Goal> goal)
{
    RCLCPP_INFO_STREAM(
        this->get_logger(),
        "Received goal pose [" << goal->pose.header.frame_id << "] "
                               << "pos=(" << goal->pose.pose.position.x << ", "
                               << goal->pose.pose.position.y << ")"
        );
    return rclcpp_action::GoalResponse::ACCEPT_AND_EXECUTE;
}

rclcpp_action::CancelResponse RobotAction::handle_cancel(
    const std::shared_ptr<rclcpp_action::ServerGoalHandle<esn_msgs::action::PickObject>> /*goal_handle*/)
{
    RCLCPP_INFO(this->get_logger(), "Cancel request received");
    return rclcpp_action::CancelResponse::ACCEPT;
}

void RobotAction::handle_accepted(
    const std::shared_ptr<rclcpp_action::ServerGoalHandle<esn_msgs::action::PickObject>> goal_handle)
{
    std::thread(&RobotAction::execute, this, goal_handle).detach();
}

void RobotAction::execute(
    const std::shared_ptr<rclcpp_action::ServerGoalHandle<esn_msgs::action::PickObject>> goal_handle)
{
    RCLCPP_INFO(this->get_logger(), "Executing goal...");

    const auto goal = goal_handle->get_goal();

    const double target_x = goal->pose.pose.position.x;
    const double target_y = goal->pose.pose.position.y;

    const double dx = target_x - this->initial_x_;
    const double dy = target_y - this->initial_y_;

    const double T = this->motion_duration_ > 1e-6 ? this->motion_duration_ : 1e-6;
    const double vx = dx / T;
    const double vy = dy / T;

    RCLCPP_INFO(this->get_logger(),
                "Computed velocity: vx=%.3f, vy=%.3f to reach in %.2fs", vx, vy, T);

    const double rate_hz = 10.0;
    rclcpp::Rate rate(rate_hz);
    const double dt = 1.0 / rate_hz;

    double t = 0.0;
    geometry_msgs::msg::Twist cmd;
    geometry_msgs::msg::Twist stop;  // zero cmd

    auto feedback = std::make_shared<esn_msgs::action::PickObject::Feedback>();
    auto result   = std::make_shared<esn_msgs::action::PickObject::Result>();

    while (t < T && rclcpp::ok())
    {
        if (goal_handle->is_canceling())
        {
            this->cmd_vel_pub_->publish(stop);

            feedback->progress = static_cast<float>(t / T);
            goal_handle->publish_feedback(feedback);

            result->success = false;
            goal_handle->canceled(result);

            RCLCPP_WARN(this->get_logger(),
                        "Goal canceled at %.0f%%", feedback->progress * 100.0);
            return;
        }

        cmd.linear.x = vx;
        cmd.linear.y = vy;
        this->cmd_vel_pub_->publish(cmd);

        feedback->progress = static_cast<float>(t / T);
        goal_handle->publish_feedback(feedback);

        rate.sleep();
        t += dt;
    }

    this->cmd_vel_pub_->publish(stop);

    if (rclcpp::ok())
    {
        result->success = true;
        goal_handle->succeed(result);
        RCLCPP_INFO(this->get_logger(),
                    "Goal succeeded, reached target pose");

        this->initial_x_ = target_x;
        this->initial_y_ = target_y;
    }
}

}
